Introduce capability levels (#355)
riscv/riscv-cheri: src/level-ext.adoc

[#section_ext_cheri_levels]
== "{cheri_levels_ext_name}" Extension for Capability Levels

{cheri_levels_ext_name} is an optional extension to {cheri_base_ext_name} that adds support for associating a level with capabilities and limiting flow of capabilities to specific memory region subsets.
This extension allows assigning a level to capabilities, which in conjunction with two new permissions allows enforcing invariants on capability propagation.
For example, this can be used to ensure that a callee cannot store a copy of the passed in argument to memory or to avoid sharing of compartment-local data between compartments.
The number of supported levels is configurable, but this specification currently only requires supporting two levels (_local_ and _global_).

=== Capability format changes

> The {cheri_levels_ext_name} adds a new `LVLBITS`-bit field to the <<section_cap_encoding,capability encoding>>, the _<<section_cap_level>>_.

Reviewer: ```suggestion
{cheri_levels_ext_name} adds a new `LVLBITS`-bit field to the <<section_cap_encoding,capability encoding>>, the _<<section_cap_level>>_.
```

[#cap_perms_encoding_levels32,width="100%",options=header,cols="^2,^1,^1,^1,^1,^1,^1,^1,^1,^2,4",align="center"]
|==============================================================================
|Bits[4:3]| R | W | C | LM | EL | SL | X | ASR | Mode^1^ | Notes
11+| *Quadrant 0: Non-capability data read/write*
11+| bit[2] - write, bit[1] - reserved (0), bit[0] - read
11+| _Reserved bits for future extensions are 0 so new permissions are not implicitly granted_
| 0   |   |   |   |   |   |   |   |   | N/A | No permissions
| 1   | ✔ |   |   |   |   |   |   |   | N/A | Data RO
| 2-3   10+| reserved
| 4   |   | ✔ |   |   |   |   |   |   | N/A | Data WO
| 5   | ✔ | ✔ |   |   |   |   |   |   | N/A | Data RW
| 6-7   10+| reserved
11+| *Quadrant 1: Executable capabilities*
11+| bit[0] - <<m_bit>> ({CAP_MODE_VALUE}-{cheri_cap_mode_name}, {INT_MODE_VALUE}-{cheri_int_mode_name})
|Bits[4:3]| R | W | C | LM | EL | SL  | X | ASR | Mode^1^ |
| 0-1   | ✔ | ✔ | ✔ | ✔  | ✔  | ∞   | ✔ |  ✔  | Mode^1^  | Execute + ASR (see <<infinite-cap>>)
| 2-3   | ✔ |   | ✔ | ✔  | ✔  | ∞   |   |     | Mode^1^  | Execute + Data & Cap RO

> | 4-5   | ✔ | ✔ | ✔ | ✔  | ✔  | ∞   | ✔ |     | Mode^1^  | Execute + Data & Cap RW

Reviewer 2: I guess by infinity you mean the maximum level i.e. the most global?

Author: Correct, I could also just use "Any" here?

Reviewer 2: I am ok with the infinity symbol

[#section_ext_cheri_levels]
== "{cheri_levels_ext_name}" Extension for Capability Levels

{cheri_levels_ext_name} is an extension to {cheri_base_ext_name} that adds support for associating a level with capabilities and limiting the flow of capabilities to specific memory region subsets.
This extension allows assigning a level to capabilities, which in conjunction with two new permissions allows enforcing invariants on capability propagation.
For example, this can be used to ensure that a callee can only write a copy of the passed-in argument capability to specific locations in memory (e.g. the callee's stack frame).
It can also be used to avoid sharing of compartment-local data (such as pointers to a stack object) between compartments.

NOTE: This specification only defines the architectural mechanics of this feature, for further information on how this can be used by software please refer to cite:[cheri-v9-spec].

> The number of supported capability levels is implementation-defined, but this specification currently only requires supporting two levels (_local_ and _global_).

Author: I don't think this is true - I tried to be generic everywhere but may have missed something.

See <<section_ext_cheri_multiple_levels>> for the mechanics when `LVLBITS>1`.

=== Capability format changes
{cheri_levels_ext_name} adds a new `LVLBITS`-bit field to the <<section_cap_encoding,capability encoding>>, the _<<section_cap_level>>_.
It also adds two new permission fields, <<el_perm>> and <<sl_perm>>.

- For `MXLEN=64` <<section_cap_encoding,capability encoding>>, the AP field is widened by `LVLBITS+1` bits (i.e. 2 bits for `LVLBITS=1`)

NOTE: The <<cap_encoding_xlen64,`MXLEN=64` capability encoding>> diagram shows the layout for `LVLBITS=1`

- For `MXLEN=32` the capability's <<cap_encoding_xlen32,AP>> field is able to encoding these permissions without increasing in size (provided that LVLBITS&#8804;2).

[#section_cap_level,reftext="capability level"]
==== Capability Level (CL)

With `LVLBITS=1`, the _Capability Level_ can hold two values: when set to 1 the capability is _global_ (in general allowing it to be stored using any authorizing capability), and when set to 0 the capability is _local_, and can only be stored by authorizing capabilities with the <<sl_perm>> set.

> Furthermore, the <<el_perm>> can be used to restrict loading of _global_ capabilities -- causing the loaded value to be _local_ instead.

Reviewer 2: ```suggestion
Furthermore, the <<el_perm>> can be used to restrict loading of _global_ capabilities -- causing the hardware to automatically set the level of loaded capabilities to _local_ instead.
```

[#section_cap_level_change]
=== Changing capability levels and permissions
While capability levels are conceptually a label on the capability rather than a permission granted by the capability, they are adjusted using the <<ACPERM>> instruction.
This avoids the need for a dedicated instruction and allows reducing the level and removing <<el_perm>> in a single instruction.

[#section_ext_cheri_multiple_levels]
=== Extending {cheri_levels_ext_name} to more than two levels
When `LVLBITS>1`, the behaviour of <<ACPERM>> can no longer use masking to adjust the <<section_cap_level>> or <<sl_perm>>, but instead must perform an integer minimum operation on those `LVLBITS`-wide fields.
The <<section_cap_level,CL>> field of the resulting capability is set to `min(rs2[CL], cs1[CL])`  (equivalent to `rs2[CL] & cs1[CL]` for `LVLBITS=1`).
Similarly, <<sl_perm>> is set to `min(rs2[SL], cs1[SL])` (equivalent to `rs2[SL] & cs1[SL]` for `LVLBITS=1`).

When storing capabilities, the <<sl_perm>> checks need to perform a `LVLBITS`-wide integer comparison instead of just testing a single bit.
Considering for an example `LVLBITS=2`:

> [grid=rows,cols="2,3,6"]

Reviewer 2: ```suggestion
[options=header,grid=rows,cols="2,3,6"]
```

| 6       | ✔ | ✔ | ✔ | ✔  | ✔  | 1     |   |     | N/A | Data & Cap RW (with store _local_)
| 7       | ✔ | ✔ | ✔ | ✔  | ✔  | 0     |   |     | N/A | Data & Cap RW (no store _local_)
|==============================================================================

[#section_cap_level_change]
=== Changing capability levels and permissions
While capability levels are conceptually a label on the capability rather than a permission granted by the capability, they are adjusted using the <<ACPERM>> instruction.
This avoids the need for a dedicated instruction and allows reducing the level and removing <<el_perm>> in a single instruction.

<<<

[#section_cap_level_summary]
=== Capability level summary table

.{cheri_levels_ext_name} `LVLBITS=1` summary table for stored capabilities
[#cap_level_store_summary,width="100%",options=header,align="center",cols="1,1,1,1,5"]
|==============================================================================
3+|Auth cap field | Data cap field | Notes
|*W*|*C*|*SL*|*CL*|
|1  | 1 | 1  | X  | Store data capability unmodified
|1  | 1 | 0  | 1  | Store data capability unmodified
|1  | 1 | 0  | 0  | Store data capability with tag cleared

> |==============================================================================

Reviewer 2: I would make this table like this instead:

|SL|CL| ~SL | Tag|
|-|-|-|-|
| 0 | 0 | 1 | 0 |
| 0 | 1 | 1 | Unchanged
| 1 | 0 | 0 | Unchanged
| 1 | 1 | 0 | Unchanged

Author: I've updated the table now and kept the ~SL as part of a comment instead which is easier to read IMO.

NOTE: if W=0 or C=0 then SL is irrelevant

.{cheri_levels_ext_name} `LVLBITS=1` summary table for loaded capabilities
[#cap_level_load_summary,width="100%",options=header,align="center",cols="1,1,1,1,1,5"]
|==============================================================================
4+|Auth cap field | Data cap field |Notes
|*R*|*C*|*EL*|*CL*| Tag            |
|1  | 1 | 1  | X  | X              |Load data capability unmodified
|1  | 1 | 0  | 1  | X              |Load data capability unmodified
|1  | 1 | 0  | 0  | 1              |Load data capability with CL=0
|1  | 1 | 0  | 0  | 0              |Load data capability unmodified

> |==============================================================================

Reviewer 2: I think this table should be like this instead:
| CL | EL | | CL | EL |
|-|-|-|-|-|
| 0 | 0 | | 0 | 0 |
| 0 | 1 | | Unchanged | Unchanged |
| 1 | 0 | | Unchanged | 0 |
| 1 | 1 | | Unchanged | Unchanged |

Author: Fixed and cleaned up table.
